File: services/go-orchestrator/pkg/metrics/metrics.go
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	// HTTP request counter
	HTTPRequestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "http_requests_total",
			Help: "Total number of HTTP requests",
		},
		[]string{"method", "endpoint", "status"},
	)

	// HTTP request duration histogram
	HTTPRequestDuration = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "http_request_duration_seconds",
			Help:    "HTTP request duration in seconds",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"method", "endpoint"},
	)

	// HTTP error counter
	HTTPErrorsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "http_requests_errors_total",
			Help: "Total number of HTTP errors",
		},
		[]string{"method", "endpoint", "status"},
	)

	// RPC request counter
	RPCRequestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "rpc_requests_total",
			Help: "Total number of RPC requests",
		},
		[]string{"method", "status"},
	)

	// RPC request duration
	RPCRequestDuration = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "rpc_request_duration_seconds",
			Help:    "RPC request duration in seconds",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"method"},
	)

	// Active workers gauge
	ActiveWorkers = promauto.NewGauge(
		prometheus.GaugeOpts{
			Name: "active_workers",
			Help: "Number of active workers connected",
		},
	)

	// Gradient aggregation counter
	GradientAggregationsTotal = promauto.NewCounter(
		prometheus.CounterOpts{
			Name: "gradient_aggregations_total",
			Help: "Total number of gradient aggregations performed",
		},
	)
)

File: services/python-ai-client/app/observability/metrics.py
"""Prometheus metrics for the Python AI service."""

from prometheus_client import Counter, Histogram, Gauge, start_http_server
import logging

logger = logging.getLogger(__name__)

# HTTP request counter
http_requests_total = Counter(
    "http_requests_total",
    "Total number of HTTP requests",
    ["method", "endpoint", "status"],
)

# HTTP request duration histogram
http_request_duration_seconds = Histogram(
    "http_request_duration_seconds",
    "HTTP request duration in seconds",
    ["method", "endpoint"],
)

# HTTP error counter
http_requests_errors_total = Counter(
    "http_requests_errors_total",
    "Total number of HTTP errors",
    ["method", "endpoint", "status"],
)

# Training metrics
training_batches_total = Counter(
    "training_batches_total", "Total number of training batches processed"
)

training_epochs_total = Counter(
    "training_epochs_total", "Total number of training epochs completed"
)

training_loss = Gauge("training_loss", "Current training loss")

training_accuracy = Gauge("training_accuracy", "Current training accuracy")

# Model metrics
model_parameters_total = Gauge(
    "model_parameters_total", "Total number of model parameters"
)

gradient_updates_total = Counter(
    "gradient_updates_total", "Total number of gradient updates sent"
)


def setup_metrics(port=8081):
    """Start the Prometheus metrics HTTP server."""
    try:
        start_http_server(port)
        logger.info(f"📊 Metrics server started on port {port}")
    except Exception as e:
        logger.error(f"❌ Failed to start metrics server: {e}")
